### PR TITLE
Adds the `plan` command

### PR DESCRIPTION
**Pull request by gsoltis**
* shows what packages would be affected by an invocation of `turbo run` with equivalent arguments
 * can use `--json` flag to produce output as a pretty-printed json array

### CODE REVIEW

File: cli/internal/plan/plan.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+	Ignore               []string
+	Cwd                  string
+	IncludeDepdendencies bool
+	NoDependents         bool
+	OutputJSON           bool
+}
+
+func (po *planOpts) ScopeOpts() *scope.Opts {
+
+	// includeDependencies := false
```
Comment (jaredpalmer) [inline cli/internal/plan/plan.go:67]: **nit:** remove comments

**Review by jaredpalmer — APPROVED** [review]
Do you think we should make this a flag on `run`? If we think about what we would need to use `--plan` to clean up e2e tests, I think we would need payload of package in scope, task ids + hashes. What do you think?

**Comment by weyert** [comment]
> Do you think we should make this a flag on `run`? If we think about what we would need to use `--plan` to clean up e2e tests, I think we would need payload of package in scope, task ids + hashes. What do you think?

I think it would make sense to add it to the `run`-command as it is related to it? I would assume otherwise you would have something?

```
turbo plan build --scope=blah --ignore=blah --since=orgin/main --json=generated-plan.json
turbo run build --scope=blah --ignore=blah --since=orgin/main
```

or is the idea to offer something like:
`turbo run --planning-file=./generated-plan.json`

File: cli/internal/plan/plan.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	return cmd.Short
+}
+
+type planOpts struct {
+	Scope                []string
+	GlobalDeps           []string
+	Since                string
+	Ignore               []string
+	Cwd                  string
+	IncludeDepdendencies bool
```
Comment (weyert) [inline cli/internal/plan/plan.go:60]:
```suggestion
	IncludeDependencies bool
```

File: cli/internal/plan/plan.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+func (po *planOpts) ScopeOpts() *scope.Opts {
+
+	// includeDependencies := false
+	// if po.IncludeDepdendencies != nil {
+	// 	includeDependencies = po.IncludeDepdendencies
+	// } else if po.Since != "" && len(po.Scope) != 0 {
+	// 	includeDependencies = true
+	// }
+	return &scope.Opts{
+		IncludeDependencies: po.IncludeDepdendencies,
```
Comment (weyert) [inline cli/internal/plan/plan.go:74]: Same here: `IncludeDepdendencies`

File: cli/internal/plan/plan.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+			depsWasSet := cmd.Flags().Changed("include-dependencies")
+			if !depsWasSet && opts.Since != "" && len(opts.Scope) != 0 {
+				opts.IncludeDepdendencies = true
+			}
+			return plan(pc.config, pc.ui, opts)
+		},
+	}
+	flags := cmd.Flags()
+	flags.StringSliceVar(&opts.Scope, "scope", nil, "Specify package(s) to act as entry points for task\nexecution. Supports globs.")
+	flags.StringSliceVar(&opts.GlobalDeps, "global-deps", nil, "Specify glob of global filesystem dependencies to\nbe hashed. Useful for .env and files in the root\ndirectory. Can be specified multiple times.")
```
Comment (weyert) [inline cli/internal/plan/plan.go:103]: Would this override the global dependencies defined in the`turbo.json` pipeline file?

**Review by weyert — COMMENTED** [review]
Looking promising

File: cli/internal/plan/plan.go (hunk 4 above)
Comment (jaredpalmer) [reply cli/internal/plan/plan.go:103]: Yes

**Comment by jaredpalmer** [comment]
@weyert serialized graph is a different matter which we are going to support. The goal here is to increase observability and unblock folks using `Ignore Command` on Vercel/Netlify and also ppl using dynamic pipelines in CircleCI and GitLab CI

**Comment by gsoltis** [comment]
@jaredpalmer  re: cleaning up e2e tests, I think we need to be able to check "packages in scope", "tasks to be run", and "what was run from cache". 

As currently implemented, `plan` gives us "packages in scope". Maybe `plan` is not the best term for that functionality, since it suggests "things that will be run", which is more like "tasks to be run". Perhaps the command here is better described as `affected`? It answers the question of "given these flags (`--since`, `--scope`, etc.), what packages are in scope?". Then a true `plan` would be "tasks to be run", a superset of "packages in scope". We don't technically need both, could even have a `--packages-only` flag to the "tasks to be run" command. What do you think, should we implement that instead?

I'm not sure I have a strong opinion on a separate command vs flag to `run`. Internally we can make the code sharing work out either way. My inclination is towards separate command because not all of the flags for `run` apply and that could lead to some documentation or parsing awkwardness, but that's not necessarily a strong argument. If there's a better DX reason either way, that should probably take precedence.

**Comment by alexandernanberg** [comment]
As a user I think I'd prefer `turbo run --dry` or an variation of that. `--dry`/`--dry-run` is quite common in other tools so might be easier to understand compared to `--plan`

**Comment by jaredpalmer** [comment]
Agree with @alexandernanberg —dry makes sense. Obvs let’s keep the abstraction needed to get the graph and scope outside of run command. These will be useful for the query commands we chatted about in our sync yesterday.

**Comment by gsoltis** [comment]
Closing in favor of #862

**Pull request closed without merging**